test/state-component-mocha.js

```diff
@@ -0,0 +1,288 @@
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'a', {a:{id:"1"}, c: "A"});
+                sendData(component, 'a', {a:{id:"2"}, c: "A"});
+                sendData(component, 'a', {a:{id:"3"}, c: "A"});
+                sendData(component, 'b', {b:{id:"2"}, c: "B"});
+            });
+        }).should.become({a: {a:{id: "2"}, c: "A"}, b: {b:{id: "2"}, c: "B"}});
+    });
+    it("map indexed data by function", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:{
+                a: {required: true, indexBy: _.property('aid')},
+                b: {required: true, indexBy: _.property('bid')}
+            },
+            onchange: function(state) {
+                handler(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'a', {aid:"1", c: "A"});
+                sendData(component, 'a', {aid:"2", c: "A"});
+                sendData(component, 'a', {aid:"3", c: "A"});
+                sendData(component, 'b', {bid:"2", c: "B"});
+            });
+        }).should.become({a: {aid: "2", c: "A"}, b: {bid: "2", c: "B"}});
+    });
+    it("map indexed data by function context", function() {
+        var handler;
+        var indexBy = function(obj) {
+            return obj[this.property];
+        };
+        return Promise.resolve({
+            inPorts:{
+                a: {required: true, indexBy: indexBy},
+                b: {required: true, indexBy: indexBy},
+                property: {ondata: function(property) {
+                    this.property = property;
+                }}
+            },
+            onchange: function(state) {
+                handler(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'property', 'id');
+                sendData(component, 'a', {id:"1", c: "A"});
+                sendData(component, 'a', {id:"2", c: "A"});
+                sendData(component, 'a', {id:"3", c: "A"});
+                sendData(component, 'b', {id:"2", c: "B"});
+            });
+        }).should.become({a: {id: "2", c: "A"}, b: {id: "2", c: "B"}});
+    });
+    function createComponent(getComponent) {
```

Comment:
createComponent and sendData are helper methods with identical code copied and pasted between promise-component-mocha.js and state-component-mocha.js.  They should be a in a common helper file that both can use so we don't proliferate copy/pasted code and have one place to maintain them.